File: services/clinic_service.go
package services

import (
	"context"

	"doctor-service/dto"
	"doctor-service/entities"
	"doctor-service/repositories"

	"github.com/google/uuid"
)

type ClinicService interface {
	AddClinic(ctx context.Context, doctorID uuid.UUID, req dto.AddClinicRequest) (*dto.ClinicResponse, error)
	GetClinics(ctx context.Context, doctorID uuid.UUID) ([]dto.ClinicResponse, error)
	DeleteClinic(ctx context.Context, clinicID, doctorID uuid.UUID) error
}

type DefaultClinicService struct {
	clinicRepo repositories.ClinicRepository
}

func NewClinicService(clinicRepo repositories.ClinicRepository) ClinicService {
	return &DefaultClinicService{clinicRepo: clinicRepo}
}

func (s *DefaultClinicService) AddClinic(ctx context.Context, doctorID uuid.UUID, req dto.AddClinicRequest) (*dto.ClinicResponse, error) {
	mode := entities.ConsultationMode(req.ConsultationMode)
	if mode == "" {
		mode = entities.ConsultationModeOffline
	}

	clinic := &entities.DoctorClinic{
		DoctorID:         doctorID,
		ClinicName:       req.ClinicName,
		Address:          req.Address,
		City:             req.City,
		State:            req.State,
		Pincode:          req.Pincode,
		Latitude:         req.Latitude,
		Longitude:        req.Longitude,
		ConsultationMode: mode,
		ConsultationFee:  req.ConsultationFee,
	}

	if err := s.clinicRepo.Create(ctx, clinic); err != nil {
		return nil, err
	}

	return &dto.ClinicResponse{
		ClinicID:         clinic.ClinicID,
		ClinicName:       clinic.ClinicName,
		Address:          clinic.Address,
		City:             clinic.City,
		State:            clinic.State,
		Pincode:          clinic.Pincode,
		ConsultationMode: string(clinic.ConsultationMode),
		ConsultationFee:  clinic.ConsultationFee,
		CreatedAt:        clinic.CreatedAt,
	}, nil
}

func (s *DefaultClinicService) GetClinics(ctx context.Context, doctorID uuid.UUID) ([]dto.ClinicResponse, error) {
	clinics, err := s.clinicRepo.FindByDoctorID(ctx, doctorID)
	if err != nil {
		return nil, err
	}

	var res []dto.ClinicResponse
	for _, c := range clinics {
		res = append(res, dto.ClinicResponse{
			ClinicID:         c.ClinicID,
			ClinicName:       c.ClinicName,
			Address:          c.Address,
			City:             c.City,
			State:            c.State,
			Pincode:          c.Pincode,
			ConsultationMode: string(c.ConsultationMode),
			ConsultationFee:  c.ConsultationFee,
			CreatedAt:        c.CreatedAt,
		})
	}
	return res, nil
}

func (s *DefaultClinicService) DeleteClinic(ctx context.Context, clinicID, doctorID uuid.UUID) error {
	return s.clinicRepo.Delete(ctx, clinicID, doctorID)
}

File: backend_api/services/clinic_service.py
from datetime import datetime
from decimal import Decimal
from entities.clinic import DoctorClinic, ConsultationMode


class ClinicService:
    def AddClinic(self, ctx, doctor_id, req):
        pass

    def GetClinics(self, ctx, doctor_id):
        pass

    def DeleteClinic(self, ctx, clinic_id, doctor_id):
        pass


class DefaultClinicService(ClinicService):
    def __init__(self, doctor_repo, clinic_repo):
        self.doctorRepo = doctor_repo
        self.clinicRepo = clinic_repo

    def _to_response(self, clinic):
        class ClinicResponse:
            pass

        resp = ClinicResponse()
        resp.clinic_id = str(clinic.clinic_id)
        resp.clinic_name = clinic.clinic_name
        resp.address = clinic.address
        resp.city = clinic.city
        resp.state = clinic.state
        resp.pincode = clinic.pincode
        resp.consultation_mode = (
            clinic.consultation_mode.value
            if hasattr(clinic.consultation_mode, "value")
            else str(clinic.consultation_mode)
        )
        resp.consultation_fee = float(clinic.consultation_fee or 0)
        resp.created_at = clinic.created_at or datetime.utcnow()
        return resp

    def AddClinic(self, ctx, doctor_id, req):
        doc = self.doctorRepo.FindByPublicID(str(doctor_id))
        if not doc:
            raise Exception("doctor not found")

        mode_raw = getattr(req, "consultation_mode", None) or "IN_PERSON"
        if mode_raw == "OFFLINE":
            mode_raw = "IN_PERSON"
        try:
            mode = ConsultationMode(mode_raw)
        except Exception:
            mode = ConsultationMode.IN_PERSON

        fee = getattr(req, "consultation_fee", None)
        if fee is None:
            fee = 0

        clinic = DoctorClinic(
            doctor_id=str(doc.id),
            clinic_name=req.clinic_name,
            address=req.address,
            city=req.city,
            state=req.state,
            pincode=req.pincode,
            latitude=getattr(req, "latitude", None),
            longitude=getattr(req, "longitude", None),
            consultation_mode=mode,
            consultation_fee=Decimal(str(fee)),
        )
        self.clinicRepo.Create(clinic)
        return self._to_response(clinic)

    def GetClinics(self, ctx, doctor_id):
        doc = self.doctorRepo.FindByPublicID(str(doctor_id))
        if not doc:
            raise Exception("doctor not found")
        clinics = self.clinicRepo.FindByDoctorID(doc.id)
        return [self._to_response(c) for c in clinics]

    def DeleteClinic(self, ctx, clinic_id, doctor_id):
        doc = self.doctorRepo.FindByPublicID(str(doctor_id))
        if not doc:
            raise Exception("doctor not found")
        return self.clinicRepo.Delete(str(clinic_id), doc.id)
